_posts/2016-01-06-angular2-taking-advantage-of-observables.md

```diff
@@ -0,0 +1,227 @@
+after the request that carries the results for *car*. This may happen because they are served by different servers. If we don't deal with such cases properly we may end up showing results for *computer* whereas the search box reads *car*.
+
+## Challenge accepted
+
+We will use the free and open wikipedia API to write a little demo.
+
+For simplicity our demo will simply consist of two files: `app.ts` and `wikipedia-service.ts`. In a real world scenario we would most likely split things further up though.
+
+Let's start with a Promise-based implementation that doesn't handle any of the described edge cases.
+
+This is what our `WikipediaService` looks like. Despite the fact that the Http/JsonP API still has some little unergonomic parts, there shouldn't be much of surprise here.
+
+{% highlight ts %}
+import {Injectable} from 'angular2/core';
+import {URLSearchParams, Jsonp} from 'angular2/http';
+
+@Injectable()
+export class WikipediaService {
+  constructor(private jsonp: Jsonp) {}
+
+  search (term: string) {
+    var search = new URLSearchParams()
+    search.set('action', 'opensearch');
+    search.set('search', term);
+    search.set('format', 'json');
+    return this.jsonp
+                .get('http://en.wikipedia.org/w/api.php?callback=JSONP_CALLBACK', { search })
+                .toPromise()
+                .then((request) => request.json()[1]);
+  }
+}
+{% endhighlight %}
+
+Basically we are injecting the `Jsonp` service to make a `GET` request against the wikipedia API with a given search term. Notice that we call `toPromise` in order to get from an `Observable<Response>` to a `Promise<Response>`. With a little bit of `then`-chaining we eventually end up with a `Promise<Array<string>>` as the return type of our `search` method.
+
+So far so good, let's take a look at the `app.ts` file that holds our `App` Component.
+
+{% highlight ts %}
+// check the plnkr for the full list of imports
+import {...} from '...';
+
+@Component({
+  selector: 'my-app',
+  template: `
+    <div>
+      <h2>Wikipedia Search</h2>
+      <input #term type="text" (keyup)="search(term.value)">
+      <ul>
+        <li *ngFor="#item of items">{{item}}</li>
+      </ul>
+    </div>
+  `
+})
+export class App {
+  items: Array<string>;
+  constructor(private wikipediaService: WikipediaService) {
+  }
+
+  search (term) {
+    this.wikipediaService.search(term)
+                         .then(items => this.items = items);
+  }
+}
+
+bootstrap(App, [WikipediaService, JSONP_PROVIDERS])
+    .catch(err => console.error(err));
+{% endhighlight %}
+
+Not much of a surprise here either. We inject our `WikipediaService` and expose it's functionality via a `search` method to the template. The template simply binds to `keyup` and calls `search(term.value)` leveraging Angular 2's awesome *template ref* feature.
```

Comment:
I think `input` event is faster than `keyup` but yeh.. this is for demoing so I'm okay with it.